executor.go

```diff
@@ -48,9 +51,160 @@ type ValidationResult struct {
+	}
+
+	return &ValidationResult{
+		Succeeded:       decision,
+		ArtifactReports: aggregatedVerifierReports,
+	}, nil
+}
+
+// aggregateVerifierReports generates and aggregates all verifier reports.
+func (e *Executor) aggregateVerifierReports(ctx context.Context, opts ValidateArtifactOptions) ([]*ValidationReport, error) {
+	var aggregatedVerifierReports []*ValidationReport
+	var err error
+
+	// TODO: Implement a worker pool to validate artifacts concurrently.
+	// TODO: Enforce check on the stack size.
+	taskStack := make([]*task, 0)
+
+	// Iterate over all stores to create tasks for each store so that further
+	// referrer artifacts will be fetched from the same store.
+	for _, store := range e.Stores {
+		taskStack = append(taskStack, &task{
+			store:    store,
+			artifact: opts.Subject,
+		})
+	}
+
+	for len(taskStack) > 0 {
+		task := taskStack[len(taskStack)-1]
+		taskStack = taskStack[:len(taskStack)-1]
+
+		taskStack, aggregatedVerifierReports, err = e.processTask(ctx, task, taskStack, aggregatedVerifierReports, opts.ReferenceTypes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate artifact %s: %w", task.artifact, err)
+		}
+	}
+
+	return aggregatedVerifierReports, nil
+}
+
+// processTask processes the task and returns the updated stack and aggregated
+// verifier reports.
+func (e *Executor) processTask(ctx context.Context, task *task, stack []*task, aggregatedVerifierReports []*ValidationReport, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	artifactRef := task.artifact
+	ref, err := registry.ParseReference(artifactRef)
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to parse artifact reference %s: %w", artifactRef, err)
+	}
+
+	artifactDesc, err := task.store.Resolve(ctx, ref.String())
+	if err != nil {
+		return stack, nil, fmt.Errorf("failed to resolve artifact reference %s: %w", ref.Reference, err)
+	}
+	ref.Reference = artifactDesc.Digest.String()
+
+	stack, artifactReports, err := e.verifyAgainstReferrers(ctx, task.store, ref, stack, referenceTypes)
+	if err != nil {
+		return stack, nil, err
+	}
+
+	// If the current task is the root task, add the artifactReports to the
+	// aggregatedVerifierReports. Otherwise, they are just artifactReports of
+	// the subject artifact.
+	if task.subjectReport != nil {
+		task.subjectReport.ArtifactReports = append(task.subjectReport.ArtifactReports, artifactReports...)
+	} else {
+		aggregatedVerifierReports = append(aggregatedVerifierReports, artifactReports...)
+	}
+	return stack, aggregatedVerifierReports, nil
+}
+
+// verifyAgainstReferrers verifies the subject artifact against all referrers in
+// the store and produces new tasks for each referrer.
+func (e *Executor) verifyAgainstReferrers(ctx context.Context, store Store, subject registry.Reference, stack []*task, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	referrers, err := store.ListReferrers(ctx, subject.String(), referenceTypes, nil)
```

Comment:
```go
	// ListReferrers returns the immediate set of supply chain artifacts for the
	// given subject reference.
	// Note: This API supports pagination. fn should be set to handle the
	//       underlying pagination.
        ListReferrers(ctx context.Context, ref string, artifactTypes []string, fn func(referrers []ocispec.Descriptor) error) ([]ocispec.Descriptor, error)
```
This method is a little strange. It supports pagination but still returns a slice of descriptors. What happens if `fn` is not `nil`? Will this method still return all the descriptors?

Reply:
if fn is passed, we can just return nil but let fn handles the pagination. Since current store implementation will handle the fn and pagination, we just pass nil to the API to facilitate the usage.

Reply:
Got it. So can we update the function comment to define the behavior when `fn` is `nil` or not `nil`

Reply:
Updated the API